V1 to v2 doc (#549)
microsoft/presidio: docs/presidio_V2.md

# Presidio Revamp (aka V2)

As of March 2021, Presidio had undergo a revamp to a new version refereed to as **V2**.

The main changes introduced in **V2** are:

1. gRPC replaced with HTTP to allow more customizable APIs and easier debugging
2. Presidio Anonymizer is Python based and pip installable.

> 3. Focus on the Analyzer and Anonymizer services. Other services are deprecated and potentially be migrated over time to **V2** with the help of the community.

Reviewer: ```suggestion
3. Focus on the Analyzer and Anonymizer services. 

Other services are deprecated and potentially be migrated over time to **V2** with the help of the community.

```

Author: not sure this should be a seperate comment

Reviewer: it's not separate comment, just a new line :)

Reviewer: but it's a nit

4. Improved documentation, samples and build flows.
5. Presidio Analyzer - removed use of templates and external recognizer store

Notes:

> The legacy V1 code base will continue to be available under branch [V1](https://github.com/microsoft/presidio/tree/V1) but will no longer be officially supported.

Reviewer: we need to add a todo to verify this link works in the future.

## API Changes

The move from gRPC to HTTP based APIs included changes to the API requests.

> 1. Moving from GRPC to HTTP - moving from structures to jsons.

Reviewer: ```suggestion
1. Change in payload - moving from structures to jsons.

```

2. Removing templates from the API - includes flattening the json.
3. Using snake_case instead of camel_case.

Bellow is a detailed outline of all the changes done to the Analyzer and Anonymizer.

### Analyzer API Changes

#### Legacy json request (gRPC)

```json
{
    "text": "My phone number is 212-555-5555",
    "AnalyzeTemplateId": "1234",
    "AnalyzeTemplate": {
        "Fields": [
            {
                "Name": "PHONE_NUMBER",
                "MinScore": "0.5"
            }
        ],
        "AllFields": true,
        "Description": "template description",
        "CreateTime": "template creation time",
        "ModifiedTime": "template modification time",
        "Language": "fr",
        "ResultsScoreThreshold": 0.5
    }
}
```

#### V2 json request (HTTP)

```json
{
    "text": "My phone number is 212-555-5555",
    "entities": ["PHONE_NUMBER"],
    "language": "en",
    "correlation_id": "213",
    "score_threshold": 0.5,
    "trace": true,
    "remove_interpretability_response": true
}
```

### Anonymizer API Changes

#### Legacy json request (gRPC)

```json

> {
>   "text": "hello world, my name is Jane Doe. My number is: 034453334",
>   "template": {
>     "description": "DEPRECATED",
>     "create_time": "DEPRECATED",
>     "modified_time": "DEPRECATED",
>     "default_transformation": {
>       "replace_value": {...},
>       "redact_value": {...},
>       "hash_value": {...},
>       "mask_value": {...},
>       "fpe_value": {...}
>     },
>     "field_type_transformations": [{
>       "fields":[{
>         "name": "FIRST_NAME",
>         "min_score": "0.2"
>       }],
>       "transfomarion": {
>         "replace_value": {...},
>       "redact_value": {...},
>       "hash_value": {...},
>       "mask_value": {...},
>       "fpe_value": {...}
>       },]
>   },
>   "analyze_results": [
>     {
>       "text": "Jane",
>       "field": {
>         "name": "FIRST_NAME",
>         "min_score": "0.5"
>       },
>       "location": {
>         "start": 24,
>         "end": 32,
>         "length": 6
>       },
>       "score": 0.8
>     }
>   ]
> }

Reviewer: ```suggestion
{
  "text": "hello world, my name is Jane Doe. My number is: 034453334",
  "template": {
    "description": "DEPRECATED",
    "create_time": "DEPRECATED",
    "modified_time": "DEPRECATED",
    "default_transformation": {
      "replace_value": {
        ...
      },
      "redact_value": {
        ...
      },
      "hash_value": {
        ...
      },
      "mask_value": {
        ...
      },
      "fpe_value": {
        ...
      }
    },
    "field_type_transformations": [
      {
        "fields": [
          {
            "name": "FIRST_NAME",
            "min_score": "0.2"
          }
        ],
        "transfomarion": {
          "replace_value": {...},
          "redact_value": {...},
          "hash_value": {...},
          "mask_value": {...},
          "fpe_value": {...}
        }
      }
    ],
    "analyze_results": [
      {
        "text": "Jane",
        "field": {
          "name": "FIRST_NAME",
          "min_score": "0.5"
        },
        "location": {
          "start": 24,
          "end": 32,
          "length": 6
        },
        "score": 0.8
      }
    ]
  }

```

Reviewer: The json was invalid, I changed the commas and brackets to fix it.

# Presidio Revamp (aka V2)

As of March 2021, Presidio had undergo a revamp to a new version refereed to as **V2**.

The main changes introduced in **V2** are:

1. gRPC replaced with HTTP to allow more customizable APIs and easier debugging
2. Presidio Anonymizer is Python based and pip installable.

> 3. Focus on the Analyzer and Anonymizer services. Other services are deprecated and potentially be migrated over time to **V2** with the help of the community.

Reviewer 2: Consider mentioning the image redactor

Author: added a note on the image-redactor

4. Improved documentation, samples and build flows.
5. Presidio Analyzer - removed use of templates and external recognizer store

Notes:
#TODO:check link update master branch update
The legacy V1 code base will continue to be available under branch [V1](https://github.com/microsoft/presidio/tree/V1) but will no longer be officially supported.

## API Changes

The move from gRPC to HTTP based APIs included changes to the API requests.

1. Change in payload - moving from structures to jsons.

2. Removing templates from the API - includes flattening the json.

> 3. Using snake_case instead of camel_case.

Reviewer 2: nit: camelCase 😊

> Bellow is a detailed outline of all the changes done to the Analyzer and Anonymizer.

Reviewer 2: Typo: Below

### Analyzer API Changes

#### Legacy json request (gRPC)

```json
{
    "text": "My phone number is 212-555-5555",
    "AnalyzeTemplateId": "1234",
    "AnalyzeTemplate": {
        "Fields": [
            {
                "Name": "PHONE_NUMBER",
                "MinScore": "0.5"
            }
        ],
        "AllFields": true,
        "Description": "template description",
        "CreateTime": "template creation time",
        "ModifiedTime": "template modification time",
        "Language": "fr",
        "ResultsScoreThreshold": 0.5
    }
}
```

#### V2 json request (HTTP)

```json
{
    "text": "My phone number is 212-555-5555",
    "entities": ["PHONE_NUMBER"],
    "language": "en",
    "correlation_id": "213",
    "score_threshold": 0.5,
    "trace": true,

>     "remove_interpretability_response": true

Reviewer 2: Whoever merges first should update this to return_decision_process 😊

Author: done
